File: src/main/java/com/example/cqrs/query/document/AccountDocument.java
package com.example.cqrs.query.document;

import jakarta.persistence.Id;
import lombok.AccessLevel;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import org.springframework.data.mongodb.core.mapping.Document;

import java.time.LocalDateTime;

/**
 * MongoDB에 저장되는 계좌 읽기 모델
 */
@AllArgsConstructor(access = AccessLevel.PRIVATE)
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@Getter
@Document(collection = "account")
public class AccountDocument {

    @Id
    private String accountId; // 계좌 ID
    private double balance;   // 잔액
    private LocalDateTime lastUpdated; // 마지막 업데이트 시간

    // factory method
    public static AccountDocument of(String accountId, double balance, LocalDateTime lastUpdated) {
        return new AccountDocument(accountId, balance, lastUpdated);
    }

    // 잔액 변경 메서드
    public void changeBalance(double newBalance) {
        this.balance = newBalance;
        this.lastUpdated = LocalDateTime.now();
    }

    // 마지막 업데이트 시간 변경 메서드
    public void changeLastUpdated(LocalDateTime eventDate) {
        this.lastUpdated = eventDate;
    }

}

File: src/main/java/com/example/cqrs/infrastructure/persistence/query/document/AccountDocument.kt
package com.example.cqrs.infrastructure.persistence.query.document

import jakarta.persistence.Id
import org.springframework.data.mongodb.core.index.Indexed
import org.springframework.data.mongodb.core.mapping.Document
import java.time.LocalDateTime

@Document(collection = "account")
class AccountDocument private constructor(
    @Id
    @Indexed(unique = true)
    val accountId: String, // 계좌

    var balance: Double,   // 잔액

    var lastUpdated: LocalDateTime // 마지막 업데이트 시간
) {

    // 잔액 변경 메서드
    fun changeBalance(newBalance: Double) {
        this.balance = newBalance
        this.lastUpdated = LocalDateTime.now()
    }

    // 마지막 업데이트 시간 변경 메서드
    fun changeLastUpdated(newLastUpdated: LocalDateTime) {
        this.lastUpdated = newLastUpdated
    }

    companion object {
        fun of(accountId: String, balance: Double, lastUpdated: LocalDateTime): AccountDocument {
            return AccountDocument(accountId, balance, lastUpdated)
        }
    }

    // MongoDB를 위한 기본 생성자 (protected 접근 수준에 해당)
    private constructor() : this("", 0.0, LocalDateTime.MIN)

}